### PR TITLE
8318562: Computational test more than 2x slower when AVX instructions are used

### PR DESCRIPTION
**Pull request by sviswa7**
This PR fixes the perf regression seen on AVX for floating point conversions.

In AVX the cvt instructions have three operands cvtxx dst, src1, src2.  Where src2 is the one being converted. The dst gets the lower bits as the converted value and upper bits (up to 128) from src1.

The C2 jit uses the cvtxx dst, dst, src2 flavor. Here the problem was due to uninitialized upper bits of the dst XMM register.
Doing an xor dst, dst  before the conversion instruction fixes the perf regression. 

Perf before the patch on UseAVX=3 platform:
ComputePI.compute_pi_dbl_flt   avgt    5   471.875 ±  0.400  ns/op
ComputePI.compute_pi_flt_dbl   avgt    5  1877.174 ±  0.557  ns/op
ComputePI.compute_pi_int_dbl   avgt    5   655.222 ± 28.082  ns/op
ComputePI.compute_pi_int_flt   avgt    5   737.178 ±  0.077  ns/op
ComputePI.compute_pi_long_dbl  avgt    5   767.364 ±  0.027  ns/op
ComputePI.compute_pi_long_flt  avgt    5   587.854 ± 10.068  ns/op

Perf after the patch on UseAVX=3 platform:
Benchmark                      Mode  Cnt    Score   Error  Units
ComputePI.compute_pi_dbl_flt   avgt    5  468.328 ± 0.141  ns/op
ComputePI.compute_pi_flt_dbl   avgt    5  435.430 ± 0.259  ns/op
ComputePI.compute_pi_int_dbl   avgt    5  424.088 ± 0.050  ns/op
ComputePI.compute_pi_int_flt   avgt    5  417.345 ± 0.207  ns/op
ComputePI.compute_pi_long_dbl  avgt    5  425.751 ± 0.006  ns/op
ComputePI.compute_pi_long_flt  avgt    5  430.199 ± 0.736  ns/op

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8318562](https://bugs.openjdk.org/browse/JDK-8318562): Computational test more than 2x slower when AVX instructions are used (**Enhancement** - P4)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16701/head:pull/16701` \
`$ git checkout pull/16701`

Update a local copy of the PR: \
`$ git checkout pull/16701` \
`$ git pull https://git.openjdk.org/jdk.git pull/16701/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16701`

View PR using the GUI difftool: \
`$ git pr show -t 16701`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16701.diff">https://git.openjdk.org/jdk/pull/16701.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16701#issuecomment-1815530798)

### CODE REVIEW

**Review by vnkozlov — COMMENTED** [review]
@sviswa7 thank you for finding the cause! I will test it locally.

**Comment by vnkozlov** [comment]
I confirmed that this change solved performance issue on machines I tested (old Broadwell and Cascade Lake CPUs).
I am submitting our regular testing for approval.

**Comment by sviswa7** [comment]
@vnkozlov Thanks a lot!

**Comment by jatin-bhateja** [comment]
Hi @sviswa7 , 

Thanks for addressing this.

For SSE versions Bits (MAXVL-1:32) of the corresponding destination register remain unchanged

For AVX Bits (127:32) of the XMM register destination are copied from corresponding bits in the first source operand.

Since all the computation in backend intially happens in logical register and only at retirement backend copies the logical register to architectural register, thus from micro architectural standpoint both the cases will result in emittion of extra micro op, in first case for merging and in second case for copying. Which is why we inject vzeroupper to save merging plenalities b/w AVX2/AVX512 and SSE transitions.

Fix looks ok to me otherwise.

File: src/hotspot/cpu/x86/x86_64.ad [hunk 1]
```diff
@@ -11089,7 +11089,7 @@ instruct convF2D_reg_mem(regD dst, memory src)
 instruct convD2F_reg_reg(regF dst, regD src)
 %{
   match(Set dst (ConvD2F src));
-
+  effect(TEMP dst);
```
Comment (merykitty) [inline src/hotspot/cpu/x86/x86_64.ad:11092]: You don't need `TEMP dst`, if `dst` is an alias of `src` then a destructive `xor` is not emitted.

**Review by merykitty — COMMENTED** [review]
For `cvtss2sd` and `cvtsd2ss`, can we branch to have AVX use `vcvtss2sd(dst, src, src)`. This removes the redundant `xor` on newer machines.

File: src/hotspot/cpu/x86/x86_64.ad (hunk 1 above)
Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/x86_64.ad:11092]: Without TEMP annotation dst and src may be aliased if src live range does not survives beyond this instruction.

**Review by vnkozlov — APPROVED** [review]
My tier1-4,xcomp,stress passed.

File: src/hotspot/cpu/x86/x86_64.ad (hunk 1 above)
Comment (sviswa7) [reply src/hotspot/cpu/x86/x86_64.ad:11092]: I had checked before submitting the PR, the cvt  xmm0, xmm0, xmm0 form was slower than xorps xmm1, xmm1 followed by cvt xmm1, xmm1, xmm0.

**Comment by merykitty** [comment]
@sviswa7 You mean `cvt xmm1, xmm0, xmm0` is slower than `xorps xmm1, xmm1; cvt xmm1, xmm1, xmm0`, right? Since `cvt xmm0, xmm0, xmm0` has self dependency so standalone its throughput will be lower than `xorps xmm1, xmm1; cvt xmm1, xmm1, xmm0` which does not.

**Pull request closed without merging**